xla/tests/collective_ops_e2e_test.cc

```diff
@@ -1340,6 +1341,282 @@ TEST_F(CollectiveOpsTestE2E, NoAsyncCollectives) {
+                   << " devices (" << test_runner().device_count()
+                   << " available)";
+    }
+
+    // Unsharded case.
+    DeviceAssignment ref_assn(/*replica_count=*/1,
+                              /*computation_count=*/1);
+    ref_assn(0, 0) = 0;
+    HloModuleConfig ref_config = GetModuleConfigForTest(/*replica_count=*/1);
+    auto ref_opts = GetDebugOptionsForTest();
+    ref_opts.set_xla_gpu_enable_triton_gemm(false);
+    ref_config.set_debug_options(ref_opts);
+    ref_config.set_num_partitions(1);
+    TF_ASSERT_OK_AND_ASSIGN(auto ref_module, ParseAndReturnVerifiedModule(
+                                                 hlo_text_ref, ref_config));
+    const int64_t num_parameters =
+        ref_module->entry_computation()->num_parameters();
+
+    auto fake_args = xla::MakeFakeArguments(ref_module.get()).value();
+    std::vector<Literal*> ref_fake_ptrs(num_parameters);
+    for (int i = 0; i < num_parameters; ++i) {
+      ref_fake_ptrs[i] = &fake_args[i];
+    }
+    TF_ASSERT_OK_AND_ASSIGN(
+        std::vector<Literal> ref_results,
+        HloTestBase::ExecuteReplicated(std::move(ref_module), ref_fake_ptrs,
+                                       /*num_replicas=*/1, &ref_assn,
+                                       /*run_hlo_passes=*/true,
+                                       /*use-threads=*/true));
+    ASSERT_EQ(ref_results.size(), 1);
+
+    // Sharded case.
+    DeviceAssignment assn(/*replica_count=*/num_replicas,
+                          /*computation_count=*/num_partitions);
+    for (int64_t i = 0; i < num_partitions; ++i) {
+      assn(0, i) = i;
+    }
+    HloModuleConfig config =
+        GetModuleConfigForTest(/*replica_count=*/num_replicas);
+    auto opts = GetDebugOptionsForTest();
+    opts.set_xla_gpu_enable_triton_gemm(false);
+    config.set_debug_options(opts);
+    config.set_num_partitions(num_partitions);
+    TF_ASSERT_OK_AND_ASSIGN(auto module,
+                            ParseAndReturnVerifiedModule(hlo_text, config));
+
+    // Gather parameter and return value sharding information.
+    std::vector<std::vector<int64_t>> param_dims(num_parameters);
+    std::vector<std::vector<int64_t>> param_dims_per_shard(num_parameters);
+    std::vector<std::vector<int64_t>> param_sharded_dims(num_parameters);
+    for (int i = 0; i < num_parameters; ++i) {
+      param_dims[i] = std::vector(module->entry_computation()
+                                      ->parameter_instruction(i)
+                                      ->shape()
+                                      .dimensions()
+                                      .begin(),
+                                  module->entry_computation()
+                                      ->parameter_instruction(i)
+                                      ->shape()
+                                      .dimensions()
+                                      .end());
+      param_dims_per_shard[i] = param_dims[i];
```

Comment:
param_dims_per_shard could be generated in a helper function. Think that would make it easier to read.

Reply:
I dont think this is resolved.